[EOC-123] List member adding changing roles permissions

alemikolo opened the pull request:
1. Private lists:
- owners can add users, remove them, and change their permission
2. Cohort's lists:
- owners can add users which don't belong to cohort
- public (shared) list - all cohort's members are displayed as a list's members, removing users from public list is disabled except of situation when user doesn't belong to cohort, owners can change permissions to all list members
- private list - owners have to manually add users (cohort members or other users), remove them, and change their permissions

![chrome-capture](https://user-images.githubusercontent.com/25374390/55550349-4ca02880-56d8-11e9-9478-a46724a34b87.gif)

marekrozmus commented on server/controllers/list.js:
```diff
@@ -456,16 +483,220 @@ const removeFromFavourites = (req, resp) => {
+      }
+      const { cohortId: cohort, isPrivate, memberIds } = doc;
+      const isCohortMember = checkIfCohortMember(cohort, userId);
+      if (isPrivate || (!isPrivate && !isCohortMember)) {
```
do you need additional usage fo `isPrivate` in second part of condition?

marekrozmus commented on client/src/common/components/Members/components/MemberDetails.jsx:
```diff
@@ -296,6 +302,8 @@ MemberDetails.propTypes = {
   isOwner: PropTypes.bool,
   match: RouterMatchPropType.isRequired,
+  isPrivate: PropTypes.bool,
+  isGuest: PropTypes.bool,
```
ordering

marekrozmus commented on client/src/common/components/Members/components/MemberDetails.jsx:
```diff
@@ -244,35 +246,39 @@ class MemberDetails extends PureComponent {
-                        >
-                          Confirm
-                        </button>
+                  {(route === Routes.COHORT ||
```
create separate methods for rendering

marekrozmus commented on client/src/modules/cohort/model/selectors.js:
```diff
@@ -41,5 +41,11 @@ export const getArchivedCohorts = createSelector(
-  _keyBy(state.cohorts.data[id].members, '_id');
+export const getMembers = createSelector(
+  getCohortDetails,
+  cohort => {
```
could be simplified to
```cohort => cohort ? _keyBy(cohort.members, '_id') : undefined```

marekrozmus commented on client/src/modules/list/model/selectors.js:
```diff
@@ -42,3 +42,12 @@ export const getArchivedLists = createSelector(
+
+export const getMembers = createSelector(
+  getList,
+  list => {
```
can be simplified to:
```list => list ? _keyBy(list.member, '_id') : undefined```

marekrozmus commented on server/common/utils/index.js:
```diff
@@ -59,23 +59,72 @@ const responseWithCohorts = (cohorts, userId) =>
+    const convertedUserId = ObjectId(userId);
+    return [...memberIds, ...ownerIds].some(id => id.equals(convertedUserId));
+  }
+  return false;
```
new line before return

marekrozmus commented on server/common/utils/index.js:
```diff
@@ -59,23 +59,72 @@ const responseWithCohorts = (cohorts, userId) =>
+      ? !data.some(member => member._id.equals(userId))
+      : ![...data.memberIds, ...data.ownerIds].some(id => id.equals(userId));
+  }
+  return true;
```
new line before return

adamklepacz commented on server/common/utils/index.js:
```diff
@@ -53,29 +54,83 @@ const responseWithItem = (item, userId) => {
+    isGuest: checkIfGuest(cohortMembers, user._doc._id)
+  }));
+
+const uniqueMembers = (cohortMembers, listMembers) =>
```
what this function does?
Maybe add some small comment above this, cause it's not very clear

alemikolo replied:
All cohort's members are members of public list. public list may contain some of them in it's ownerIds but additionally it can contain guestIds which are not cohort's. So we have to merge those arrays (to have all public list members) and remove duplicated user's ids.

adamklepacz replied:
add some describing comment above this function

adamklepacz commented on server/controllers/list.js:
```diff
@@ -456,16 +483,220 @@ const removeFromFavourites = (req, resp) => {
+  const { userId } = req.body;
+  const {
+    user: { _id: ownerId }
+  } = req;
```
add break line here

adamklepacz commented on server/controllers/list.js:
```diff
@@ -456,16 +483,220 @@ const removeFromFavourites = (req, resp) => {
+      if (isPrivate || !isCohortMember) {
+        doc.memberIds.push(userId);
+      }
+      doc.ownerIds.splice(ownerIds.indexOf(userId), 1);
```
what this does?

alemikolo replied:
remove user id from ownerIds

adamklepacz commented on server/controllers/list.js:
```diff
@@ -456,16 +483,220 @@ const removeFromFavourites = (req, resp) => {
+
+const addMember = (req, resp) => {
+  const {
+    user: { _id: currentUser }
```
currentUserId

adamklepacz commented on server/controllers/list.js:
```diff
@@ -456,16 +483,220 @@ const removeFromFavourites = (req, resp) => {
+        .then(doc => {
+          if (doc) {
+            const data = { avatarUrl, displayName, newMemberId };
+            const dataToSend = responseWithListMember(data, ownerIds, cohort);
```
add line break

adamklepacz commented on server/controllers/list.js:
```diff
@@ -456,16 +483,220 @@ const removeFromFavourites = (req, resp) => {
+    })
+    .catch(err => {
+      if (err instanceof NotFoundException) {
+        const { status, message } = err;
```
add line break

marekrozmus commented on client/src/common/components/Members/components/MemberDetails.jsx:
```diff
@@ -108,7 +109,7 @@ class MemberDetails extends PureComponent {
   };
 
-  handleCohortMemberRoleChange = role => {
+  handleCohortMemberRoleChange = checkedRole => {
```
maybe better name would be `selectedRole`?

marekrozmus commented on server/common/utils/index.js:
```diff
@@ -53,29 +54,89 @@ const responseWithItem = (item, userId) => {
+
+const checkIfGuest = (data, userId) => {
+  if (data) {
+    return data.constructor === Array
```
please use: https://lodash.com/docs/4.17.11#isArray

alemikolo commented:
I fixed addMember controllers.
- I added throwing an error instead of returning response in catch methods, because in 'return' case promise wasn't rejected and the code went on.
- I changed a little bit cohort's addMember controller to fix following issue: when user was guest member in cohort's public list and was added as a member to this cohort, them still was stored in memberIds array of this list. Now such user is removed from public list's memberIds array but is still displayd as member of this list because is a cohort member now.

adamklepacz commented on client/src/common/components/Members/components/MemberDetails.jsx:
```diff
@@ -168,115 +269,31 @@ class MemberDetails extends PureComponent {
-            </header>
-            {isCurrentOwner && userId !== currentUserId && (
+            {this.renderHeader()}
+            {isCurrentUserAnOwner && userId !== currentUserId && (
```
please separate this also to renderSettings() method or something

adamklepacz commented on client/src/common/components/Members/components/MemberDetails.jsx:
```diff
@@ -168,115 +269,31 @@ class MemberDetails extends PureComponent {
               </Fragment>
             )}
-            {isCurrentOwner && userId === currentUserId && (
+            {isCurrentUserAnOwner && userId === currentUserId && (
```
extract this to renderMessage or renderNotice() method

adamklepacz commented on client/src/common/components/Members/index.jsx:
```diff
@@ -64,25 +76,26 @@ class MembersBox extends PureComponent {
             )}
           </li>
-          {_map(users, user => (
+          {_map(members, member => (
```
I know that was mine code, but could you please seprate this to renderMembersList() method??
woudl much more cleaner

adamklepacz commented on client/src/modules/cohort/index.jsx:
```diff
@@ -217,8 +218,9 @@ class Cohort extends PureComponent {
-                isCurrentOwner={this.checkIfOwner()}
+                isCurrentUserAnOwner={this.checkIfOwner()}
                 route={Routes.COHORT}
+                members={members}
```
sort props

adamklepacz commented on client/src/modules/list/index.jsx:
```diff
@@ -141,6 +149,21 @@ class List extends Component {
+                  isCurrentUserAnOwner={this.checkIfOwner()}
+                  route={Routes.LIST}
+                  members={members}
+                  isPrivate={isPrivate}
```
sort props

adamklepacz commented on server/controllers/list.js:
```diff
@@ -456,16 +494,232 @@ const removeFromFavourites = (req, resp) => {
+      }
+
+      const { cohortId: cohort, isPrivate, memberIds } = doc;
+      const isCohortMember = checkIfCohortMember(cohort, userId);
```
you can make it as :
`isNotCohortMember = !checkIfCohortMember(cohort, userId);`

adamklepacz commented on server/controllers/list.js:
```diff
@@ -456,16 +494,232 @@ const removeFromFavourites = (req, resp) => {
+      const { cohortId: cohort, isPrivate, memberIds } = doc;
+      const isCohortMember = checkIfCohortMember(cohort, userId);
+
+      if (isPrivate || !isCohortMember) {
```
and here you'll have` if(isPrivate || isNotCohortMember) `

adamklepacz commented on server/controllers/list.js:
```diff
@@ -456,16 +494,232 @@ const removeFromFavourites = (req, resp) => {
+      doc.ownerIds.push(userId);
+      return doc.save();
+    })
+    .then(() => {
```
unecessary braces

adamklepacz commented on server/controllers/list.js:
```diff
@@ -456,16 +494,232 @@ const removeFromFavourites = (req, resp) => {
+      }
+
+      const { cohortId: cohort, isPrivate, ownerIds } = doc;
+      const isCohortMember = checkIfCohortMember(cohort, userId);
```
`const isNotCohortMember = !checkIfCohortMember(<params>)`

adamklepacz commented on server/controllers/list.js:
```diff
@@ -456,16 +494,232 @@ const removeFromFavourites = (req, resp) => {
+      const { cohortId: cohort, isPrivate, ownerIds } = doc;
+      const isCohortMember = checkIfCohortMember(cohort, userId);
+
+      if (isPrivate || !isCohortMember) {
```
`if(isPrivate || isNotCohortMember)`

adamklepacz commented on server/controllers/list.js:
```diff
@@ -456,16 +494,232 @@ const removeFromFavourites = (req, resp) => {
+      doc.ownerIds.splice(ownerIds.indexOf(userId), 1);
+      return doc.save();
+    })
+    .then(() => {
```
unecessary braces

adamklepacz commented on server/controllers/list.js:
```diff
@@ -456,16 +494,232 @@ const removeFromFavourites = (req, resp) => {
+            .status(400)
+            .send({ message: 'User is already a member.' });
+        })
+        .catch(() => {
```
unecessary braces

alemikolo replied:
throw has to be inside {} braces

adamklepacz commented on client/src/common/components/Members/components/MemberDetails.jsx:
```diff
@@ -131,24 +132,124 @@ class MemberDetails extends PureComponent {
+  };
+
+  renderRemoveOption = () => {
+    const { isConfirmVisible } = this.state;
```
isConfirmationVisible

adamklepacz commented on client/src/modules/list/index.jsx:
```diff
@@ -169,16 +192,25 @@ List.propTypes = {
+    match: {
+      params: { id }
+    }
+  } = ownProps;
```
add line break
